tests/e2e/common/diagnosticcollector.go

```diff
@@ -0,0 +1,564 @@
+// diagnostic collector.
+func (d *DiagnosticCollector) ParseObjects(objects string) {
+	for _, obj := range strings.Split(objects, ",") {
+		switch obj {
+		case Nodes:
+			d.Core = append(d.Core, Nodes)
+		case Pods:
+			d.Core = append(d.Core, Pods)
+		case Namespaces:
+			d.Core = append(d.Core, Namespaces)
+		case Deployments:
+			d.Apps = append(d.Apps, Deployments)
+		case DaemonSets:
+			d.Apps = append(d.Apps, DaemonSets)
+		case Jobs:
+			d.Batch = append(d.Batch, Jobs)
+		case NodeFeature:
+			d.Extensions = append(d.Extensions, NodeFeature)
+		case NodeFeatureRule:
+			d.Extensions = append(d.Extensions, NodeFeatureRule)
+		}
+	}
+}
+
+func (d *DiagnosticCollector) RunDiagnostics(ctx context.Context) error {
+	defer d.Close()
+
+	// Collect the core objects
+	for i := range d.Core {
+		switch d.Core[i] {
+		case Nodes:
+			d.Add(d.Nodes)
+		case Pods:
+			d.Add(d.Pods)
+		case Namespaces:
+			d.Add(d.Namespaces)
+		}
+	}
+
+	// Collect the apps objects
+	for i := range d.Apps {
+		switch d.Apps[i] {
+		case Deployments:
+			d.Add(d.Deployments)
+		case DaemonSets:
+			d.Add(d.DaemonSets)
+		}
+	}
+
+	// Collect the batch objects
+	if len(d.Batch) != 0 {
+		// For now we only support jobs
+		d.Add(d.Jobs)
+	}
+
+	// Collect the extensions
+	for i := range d.Extensions {
+		d.AddExtension(d.Extensions[i])
+	}
+
+	if len(d.collectors) == 0 {
+		return fmt.Errorf("no collectors specified")
+	}
+
+	// Run the collectors
+	for _, f := range d.collectors {
+		if err := f(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Nodes is a collectorFunc that collects node objects
+func (d *DiagnosticCollector) Nodes(ctx context.Context) error {
+	nodes, err := d.clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		if _, lerr := d.errLogFile.WriteString(fmt.Sprintf("Error getting nodes: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return err
+	}
+
+	nodesFile, err := os.Create(filepath.Join(d.artifactDir, "nodes.yaml"))
+	if err != nil {
+		if _, lerr := d.errLogFile.WriteString(fmt.Sprintf("Error creating nodes.yaml: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return err
+	}
+	defer nodesFile.Close()
+
+	data, err := yaml.Marshal(nodes)
+	if err != nil {
+		if _, lerr := d.errLogFile.WriteString(fmt.Sprintf("Error marshalling nodes: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return err
+	}
+
+	_, err = nodesFile.Write(data)
+	if err != nil {
+		if _, lerr := d.errLogFile.WriteString(fmt.Sprintf("Error writing nodes.yaml: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return err
+	}
+
+	return nil
+}
+
+// Namespace is a collectorFunc that collects namespace objects
+func (d *DiagnosticCollector) Namespaces(ctx context.Context) error {
```

Comment:
All these `collectorFunc`  function all seem to follow a similar pattern:
1. Fetch objects
2. Create file
3. Marshal objects to file

Can this be factored out into common functions?

Also, since you use `collectorFunc` here is this actually an interface?

Reply:
Part 1 of the question: got it. will address
Part 2: `collectorFunc` is a type so we can then call funcs just by name

Reply:
As I mentioned offline. I think what I'm seeing here is that if whe introduce the following interface:
```
type Collector interface {
    Collect(Context) error
}
```
then each of the `collectorFunc`s is just an implementation of such an interface. Does it make sense to implement each of these `Collectors` (e.g. for Namespaces) as a separate concrete type that implements the interface? The top-level `DiagnosticsCollector` would then define a slice of these collectors -- and we could also nest them futher if required. One place where this may be reasonable is Pod logs, for example.